api/rosetta/data_block_integration_test.go

```diff
@@ -0,0 +1,322 @@
+				wantStatusCode: http.StatusUnprocessableEntity,
+				wantHandlerErr: assert.Error,
+			},
+		*/
+
+		/* TODO: when the errors we return get solidifed, add test cases to verify the errors
+		=> https://github.com/optakt/flow-dps/issues/34
+		*/
+
+	}
+
+	for _, test := range tests {
+
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+
+			t.Parallel()
+
+			enc, err := json.Marshal(test.request)
+			require.NoError(t, err)
+
+			req := httptest.NewRequest(http.MethodPost, "/block", bytes.NewReader(enc))
+			req.Header.Set(echo.HeaderContentType, echo.MIMEApplicationJSON)
+
+			rec := httptest.NewRecorder()
+
+			ctx := echo.New().NewContext(req, rec)
+
+			err = api.Block(ctx)
+			test.wantHandlerErr(t, err)
+
+			if test.wantStatusCode != http.StatusOK {
+
+				e, ok := err.(*echo.HTTPError)
+				require.True(t, ok)
+				assert.Equal(t, test.wantStatusCode, e.Code)
+
+				// nothing more to do, response validation should only be done for '200 OK' responses
+				return
+			}
+
+			assert.Equal(t, test.wantStatusCode, rec.Result().StatusCode)
+
+			// unpack response
+			var blockResponse rosetta.BlockResponse
+			require.NoError(t, json.Unmarshal(rec.Body.Bytes(), &blockResponse))
+
+			if assert.NotNil(t, blockResponse.Block) {
+				// verify that we got the data for the requested block
+				assert.Equal(t, test.request.BlockID.Index, blockResponse.Block.ID.Index)
+				assert.Equal(t, test.request.BlockID.Hash, blockResponse.Block.ID.Hash)
+
+				// verify the parent block index is correct
+				assert.Equal(t, test.request.BlockID.Index-1, blockResponse.Block.ParentID.Index)
+				assert.Equal(t, test.wantParentHash, blockResponse.Block.ParentID.Hash)
+
+				assert.Equal(t, test.wantTimestamp, blockResponse.Block.Timestamp)
+
+				if test.transactionValidator != nil {
+
+					if assert.GreaterOrEqual(t, len(blockResponse.Block.Transactions), 1) {
+						test.transactionValidator(blockResponse.Block.Transactions[0])
+					}
+				}
+			}
+		})
+	}
+}
+
+// blockRequest generates a BlockRequest with the specified parameters.
+func blockRequest(height uint64, hash string) rosetta.BlockRequest {
+
+	return rosetta.BlockRequest{
+		NetworkID: defaultNetworkID(),
+		BlockID: identifier.Block{
+			Index: height,
+			Hash:  hash,
+		},
+	}
+}
+
+func validateSingleTransfer(t *testing.T, hash string, from string, to string, amount int64) transactionValidationFn {
+
+	t.Helper()
+
+	return func(tx *object.Transaction) {
+
+		assert.Equal(t, tx.ID.Hash, hash)
+		assert.Equal(t, len(tx.Operations), 2)
+
+		relatedOperations := make(map[uint]uint)
+
+		for _, op := range tx.Operations {
+
+			// save related operation IDs in a map so we can cross reference them
+			// TODO: check - is this special-casing the test too much perhaps?
+			// operation can have multiple related operations,
+			// but also this function does state that it's verifying a single transfer
+
+			if assert.Len(t, op.RelatedIDs, 1) {
```

Comment:
Operations can have multiple related operations (at least the type indicates that, since it's a slice of related operations).

This function does imply that we're validating a single transfer. But I also do feel it's perhaps too special cased. Perhaps it would be better to create a map of `operationID` => `array of related operation IDs`, and just verify that the related operations `contain` each others IDs. Thoughts?

Reply:
No, it's fine to be special-cased, that's what we will have 99% of the time on-chain anyway.